### DISCUSSION TITLE
Add a new field type for coding.

### DISCUSSION THREAD

**Discussion by harthouse** [ExpressionEngine/ExpressionEngine#3642] [Feature Requests]
### Suggest an idea for ExpressionEngine

**Is your feature request related to a problem? Please describe.**
Our web editors sometimes need to add custom CSS or markup to an entry. The current solution is to create a custom field with Textarea; however, having code formatting available would be beneficial for efficient code editing.

**Describe the solution you'd like**
Option 1: we could add a new field type specifically for coding
Option 2: we could add an option under Textarea that allows code formatting

- [ ] I am capable and would like to work on implementation of this feature if it is considered.
![cleanshot_2023-07-28_at_12 51 08_2x_720](https://github.com/ExpressionEngine/ExpressionEngine/assets/128098439/efa6cadc-7e91-4998-ba5e-4a167e670861)

**Comment by dougblackjr** [comment]
I have definitely needed something like this a handful of times. Whether to display code on a page, or two implement code (i.e. custom, page specific styles)

**Comment by ttmwebdesign** [comment]
YES. 100% YES. Having to trust Wygwam's 'source' view to not munge something irrevocably, depending on what's in it, is tricky at best. Some things would be better off in a code field to be untouched by a typical text editor (wygwam, rte) that thinks it knows better, and sometimes plain text fields just aren't the right fit.

**Comment by swierczek** [comment]
We've actually built something like this in the past using https://codemirror.net/. I'll check with the team about open sourcing it as a starting point, if that would be beneficial. Here are some screenshots as a preview:

Field settings:
<img width="391" alt="image" src="https://github.com/ExpressionEngine/ExpressionEngine/assets/2423727/2320bb55-f63f-4064-983c-f399fd9b9747">

Field in the CMS:
<img width="1087" alt="image" src="https://github.com/ExpressionEngine/ExpressionEngine/assets/2423727/207eb95f-8a90-4845-bc87-0c8a5aa1120c">

Includes various syntax highlighters:
<img width="239" alt="image" src="https://github.com/ExpressionEngine/ExpressionEngine/assets/2423727/0841e9eb-154a-41f3-b460-cd275d70633a">

Front-end example:
The first example is in Bash has line numbers disabled and copy ability disabled.
The second example is in Swift and has line numbers enabled and copy ability enabled:
<img width="730" alt="image" src="https://github.com/ExpressionEngine/ExpressionEngine/assets/2423727/c1490997-4eef-4ee7-b7f8-3d50585b9ffb">

**Reply by ttmwebdesign** [reply]
I'm not sure but this may be different than the ask? I may be wrong but I read the original ask as wanting to have a field where one can add code, and it be included in the HTML _as code_, to work as code as if it were in a template? Not necessarily to be visible on the front end like above. Although your example is certainly a very nifty option as well!

I have in mind a perfect example where code-in-the-editor self destructs when an entry is saved because the editor, in this case Wygwam, thinks all the excess should be stripped out and only leave the HTML tags. In the case I ran into, it's a combination of HTML+Javascript to run some complicated calculators (think loan calculator). It's an inherited site that we upgraded, so I'm not sure if at one point in old EE/Wygwam it was fine with leaving the code foo as is, but it doesn't anymore so the entries have to be totally hands-off until a solution can be sorted out that won't destroy the code on save. It's possible a plain text field may work for it, but that's for a 'future to do' that hasn't been discussed yet. But a code field that doesn't strip anything or do any preprocessing and outputs it exactly as it's entered so the browser renders it as such would be much preferred.
